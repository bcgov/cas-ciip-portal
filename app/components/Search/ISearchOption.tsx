export interface ISearchOption {
  toUrl?: (string) => any;
  columnName: string;
  title: string;
  isSearchEnabled: boolean;
  isSortEnabled: boolean;
  removeSearchHeader?: boolean;
  searchOptionValues?: {display: string; value: any}[];
}
