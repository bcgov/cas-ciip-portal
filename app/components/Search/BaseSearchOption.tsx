import {ISearchOption} from './ISearchOption';

export abstract class BaseSearchOption<T> implements ISearchOption {
  constructor(display, column) {
    this.title = display;
    this.columnName = column;
  }

  columnName = '';
  title = '';

  abstract parseValue?: (string) => T;

  isSearchEnabled = true;
  isSortEnabled = true;
  removeSearchHeader = false;
}
