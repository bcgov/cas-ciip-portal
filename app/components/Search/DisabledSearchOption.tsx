import {ISearchOption} from './ISearchOption';

export class DisabledSearchOption implements ISearchOption {
  constructor(title) {
    this.title = title;
  }
  columnName: string;
  title: string;
  isSearchEnabled = false;
}
