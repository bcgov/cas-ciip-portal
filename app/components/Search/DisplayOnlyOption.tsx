import {ISearchOption} from './ISearchOption';

export class DisplayOnlyOption implements ISearchOption {
  constructor(title) {
    this.title = title;
  }
  columnName: string;
  title: string;
  isSearchEnabled = false;
  isSortEnabled = false;
}
