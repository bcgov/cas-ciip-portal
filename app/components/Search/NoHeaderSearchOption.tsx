import {ISearchOption} from './ISearchOption';

// This can be used as a UI trick to save space for the search/reset buttons
export const NoHeaderSearchOption: ISearchOption = {
  title: null,
  columnName: null,
  isSearchEnabled: false,
  removeSearchHeader: true,
  parse: null
};
