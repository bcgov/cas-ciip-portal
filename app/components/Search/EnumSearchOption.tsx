import {BaseSearchOption} from './BaseSearchOption';

export class EnumSearchOption<T> extends BaseSearchOption<string> {
  enumValues: T[];

  constructor(display, column, enumValues: T[]) {
    super(display, column);
    this.enumValues = enumValues;
    this.searchOptionValues = enumValues.map((val) => {
      return {display: val, value: val};
    });
  }

  toUrl? = (val: string) => {
    return val;
  };

  searchOptionValues: {display: T; value: T}[];
}
