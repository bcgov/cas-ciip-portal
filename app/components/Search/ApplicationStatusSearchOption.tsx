import React from 'react';
import {CiipApplicationRevisionStatus} from 'FacilitiesRowItemContainer_facilityApplication.graphql';
import {EnumSearchOption} from './EnumSearchOption';
import {Form} from 'react-bootstrap';
import {getUserFriendlyStatusLabel} from 'lib/text-transforms';
import {SearchOptionComponent} from './ISearchOption';

export class ApplicationStatusSearchOption extends EnumSearchOption<
  CiipApplicationRevisionStatus | 'NOT_STARTED'
> {
  constructor(displayName: string, argName: string, nullValueArgName: string) {
    super(displayName, argName, [
      'NOT_STARTED',
      'DRAFT',
      'SUBMITTED',
      'REQUESTED_CHANGES',
      'APPROVED',
      'REJECTED'
    ]);
    this._nullValueArgName = nullValueArgName;
  }
  _nullValueArgName: string;

  Component: SearchOptionComponent = ({onChange, filterArgs}) => {
    const handleChange = (e) => {
      const {value} = e.target;
      if (value === 'NOT_STARTED') {
        onChange(true, this._nullValueArgName);
        onChange(undefined, this.columnName);
      } else {
        onChange(undefined, this._nullValueArgName);
        onChange(value, this.columnName, this.toUrl);
      }
    };

    return (
      <td>
        <Form.Control
          as="select"
          name={this.columnName}
          value={
            (filterArgs[this.columnName] as string) ??
            (filterArgs[this._nullValueArgName] ? 'NOT_STARTED' : '')
          }
          onChange={handleChange}
          aria-label="Filter by application status"
        >
          <option key="no-filter" value="">
            ...
          </option>
          {this.searchOptionValues.map((kvp) => (
            <option key={kvp.display} value={kvp.value}>
              {getUserFriendlyStatusLabel(kvp.display)}
            </option>
          ))}
        </Form.Control>
      </td>
    );
  };
}
