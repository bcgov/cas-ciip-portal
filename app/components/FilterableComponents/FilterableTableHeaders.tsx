import React, {useState, useEffect} from 'react';
import {Button, Form} from 'react-bootstrap';
import {useRouter} from 'next/router';
import {ISearchProps} from 'components/Search/SearchProps';
import {getUserFriendlyStatusLabel} from 'lib/text-transforms';

const NONE_VALUES = [null, undefined];

const FilterableTableHeaders: React.FunctionComponent<ISearchProps> = (
  props
) => {
  const router = useRouter();

  const [searchFilters, setSearchFilters] = useState({});

  const handleFilterChange = (value, column, parseMethod) => {
    setSearchFilters({
      ...searchFilters,
      [column]: parseMethod ? parseMethod(value) : value
    });
  };

  const applySearch = (searchData) => {
    const newQuery = {};
    props.searchOptions.forEach((option) => {
      const column = option.columnName;

      if (option.isSearchEnabled) {
        if (!NONE_VALUES.includes(searchData[column]))
          newQuery[column] = searchData[column];
        else delete newQuery[column];
      }
    });

    const queryString = JSON.stringify(newQuery);

    const url = {
      pathname: router.pathname,
      query: {
        ...router.query,
        relayVars: queryString
      }
    };

    router.replace(url, url, {shallow: true});
  };

  const clearForm = () => {
    setSearchFilters({});
    applySearch({});
  };

  // functional equivalent to componentDidMount()
  useEffect(() => {
    const parsedUrl = router.query.relayVars
      ? JSON.parse(String(router.query.relayVars))
      : {};
    Object.keys(parsedUrl).forEach((key) => {
      if (NONE_VALUES.includes(searchFilters[key]))
        searchFilters[key] = parsedUrl[key];
    });
  }, []);

  return (
    <tr>
      {props.searchOptions.map((option) => {
        const column = option.columnName;
        const key = option.columnName + option.title;
        const initialValue = searchFilters[column] ?? '';

        if (option.isSearchEnabled) {
          if (!option.searchOptionValues) {
            return (
              <td key={key}>
                <Form.Control
                  placeholder="Search"
                  name={column}
                  value={initialValue}
                  onChange={(evt) =>
                    handleFilterChange(evt.target.value, column, option.toUrl)
                  }
                />
              </td>
            );
          }

          return (
            <td key={key}>
              <Form.Control
                as="select"
                placeholder="Search"
                name={column}
                value={initialValue}
                onChange={(evt) =>
                  handleFilterChange(evt.target.value, column, option.toUrl)
                }
              >
                <option key={column + '-placeholder'} value="">
                  ...
                </option>
                {option.searchOptionValues.map((kvp) => (
                  <option key={column + '-' + kvp.display} value={kvp.value}>
                    {getUserFriendlyStatusLabel(kvp.display)}
                  </option>
                ))}
              </Form.Control>
            </td>
          );
        }
        if (option.removeSearchHeader) return;
        return <td key={key} />;
      })}
      <td className="flex">
        <Button variant="outline-secondary" onClick={clearForm}>
          Clear
        </Button>
        <Button
          style={{marginLeft: '5px'}}
          variant="primary"
          onClick={() => applySearch(searchFilters)}
        >
          Apply
        </Button>
      </td>
      <style jsx>{`
        .flex {
          display: flex;
        }
      `}</style>
    </tr>
  );
};

export default FilterableTableHeaders;
