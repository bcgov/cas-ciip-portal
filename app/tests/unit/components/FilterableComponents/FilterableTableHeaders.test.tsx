import React from 'react';
import FilterableTableHeaders from 'components/FilterableComponents/FilterableTableHeaders';
import {shallow} from 'enzyme';
import {ISearchOption} from 'components/Search/ISearchOption';
import * as nextRouter from 'next/router';

const createMockSearchOption: (name: string) => ISearchOption = (name) => {
  return {
    columnName: name,
    title: name,
    isSearchEnabled: true,
    isSortEnabled: true
  };
};

nextRouter.useRouter = jest.fn();
nextRouter.useRouter.mockImplementation(() => ({
  route: '/',
  query: {
    relayVars: {},
    pageVars: {}
  }
}));

describe('The filterable table headers component', () => {
  it('renders search and reset buttons', () => {
    const rendered = shallow(<FilterableTableHeaders searchOptions={[]} />);
    expect(rendered).toMatchSnapshot();
    expect(
      rendered
        .find('Button')
        .filterWhere((node) => node.props().variant === 'outline-secondary')
        .first()
        .text()
    ).toBe('Clear');
    expect(
      rendered
        .find('Button')
        .filterWhere((node) => node.props().variant === 'primary')
        .first()
        .text()
    ).toBe('Apply');
  });

  it('renders as many td elements as search options, plus one for the search buttons', () => {
    const searchOptions: ISearchOption[] = [
      createMockSearchOption('test'),
      createMockSearchOption('test2')
    ];
    const rendered = shallow(
      <FilterableTableHeaders searchOptions={searchOptions} />
    );

    expect(rendered).toMatchSnapshot();
    expect(rendered.find('td').length).toBe(3);
  });

  it('only renders elements that are searchable', () => {
    const searchOptions = [
      {...createMockSearchOption('test'), isSearchEnabled: false},
      {...createMockSearchOption('test2'), isSearchEnabled: false},
      {...createMockSearchOption('test3'), isSearchEnabled: true}
    ];
    const rendered = shallow(
      <FilterableTableHeaders searchOptions={searchOptions} />
    );

    expect(rendered).toMatchSnapshot();
    expect(rendered.find('FormControl').length).toBe(1);
    expect(rendered.find('FormControl').first().props().name).toBe('test3');
  });

  it('renders a text input if values are not provided', () => {
    const searchOptions = [createMockSearchOption('test')];
    const rendered = shallow(
      <FilterableTableHeaders searchOptions={searchOptions} />
    );

    expect(rendered).toMatchSnapshot();
    expect(
      rendered
        .find('FormControl')
        .filterWhere((formControl) => formControl.props().as === undefined)
        .length
    ).toBe(1);
    expect(
      rendered
        .find('FormControl')
        .filterWhere((formControl) => formControl.props().as === 'select')
        .length
    ).toBe(0);
  });

  it('renders a option input if values are provided', () => {
    const searchOptions: ISearchOption[] = [
      {
        ...createMockSearchOption('test'),
        searchOptionValues: [
          {
            display: 'option1',
            value: 'optionvalue1'
          },
          {
            display: 'option2',
            value: 'optionvalue2'
          }
        ]
      }
    ];
    const rendered = shallow(
      <FilterableTableHeaders searchOptions={searchOptions} />
    );

    expect(rendered).toMatchSnapshot();
    expect(
      rendered
        .find('FormControl')
        .filterWhere((formControl) => formControl.props().as === undefined)
        .length
    ).toBe(0);
    expect(
      rendered
        .find('FormControl')
        .filterWhere((formControl) => formControl.props().as === 'select')
        .length
    ).toBe(1);

    expect(rendered.find('option').length).toBe(3);
    expect(rendered.find('option').at(0).text()).toContain('...');
    expect(rendered.find('option').at(1).text()).toContain('Option1');
    expect(rendered.find('option').at(2).text()).toContain('Option2');
  });
});
