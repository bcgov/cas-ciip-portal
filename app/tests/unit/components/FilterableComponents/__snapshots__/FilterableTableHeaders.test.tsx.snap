// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The filterable table headers component only renders elements that are searchable 1`] = `
<tr>
  <td
    key="testtest"
  />
  <td
    key="test2test2"
  />
  <td
    key="test3test3"
  >
    <FormControl
      name="test3"
      onChange={[Function]}
      placeholder="Search"
      value=""
    />
  </td>
  <td>
    <ButtonGroup
      role="group"
      toggle={false}
      vertical={false}
    >
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="success"
      >
        Search
      </Button>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="danger"
      >
        Reset
      </Button>
    </ButtonGroup>
  </td>
</tr>
`;

exports[`The filterable table headers component renders a option input if values are provided 1`] = `
<tr>
  <td
    key="testtest"
  >
    <FormControl
      as="select"
      name="test"
      onChange={[Function]}
      placeholder="Search"
      value=""
    >
      <option
        key="test-placeholder"
        value=""
      >
        ...
      </option>
      <option
        key="test-option1"
        value="optionvalue1"
      >
        option1
      </option>
      <option
        key="test-option2"
        value="optionvalue2"
      >
        option2
      </option>
    </FormControl>
  </td>
  <td>
    <ButtonGroup
      role="group"
      toggle={false}
      vertical={false}
    >
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="success"
      >
        Search
      </Button>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="danger"
      >
        Reset
      </Button>
    </ButtonGroup>
  </td>
</tr>
`;

exports[`The filterable table headers component renders a text input if values are not provided 1`] = `
<tr>
  <td
    key="testtest"
  >
    <FormControl
      name="test"
      onChange={[Function]}
      placeholder="Search"
      value=""
    />
  </td>
  <td>
    <ButtonGroup
      role="group"
      toggle={false}
      vertical={false}
    >
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="success"
      >
        Search
      </Button>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="danger"
      >
        Reset
      </Button>
    </ButtonGroup>
  </td>
</tr>
`;

exports[`The filterable table headers component renders as many td elements as search options, plus one for the search buttons 1`] = `
<tr>
  <td
    key="testtest"
  >
    <FormControl
      name="test"
      onChange={[Function]}
      placeholder="Search"
      value=""
    />
  </td>
  <td
    key="test2test2"
  >
    <FormControl
      name="test2"
      onChange={[Function]}
      placeholder="Search"
      value=""
    />
  </td>
  <td>
    <ButtonGroup
      role="group"
      toggle={false}
      vertical={false}
    >
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="success"
      >
        Search
      </Button>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="danger"
      >
        Reset
      </Button>
    </ButtonGroup>
  </td>
</tr>
`;

exports[`The filterable table headers component renders search and reset buttons 1`] = `
<tr>
  <td>
    <ButtonGroup
      role="group"
      toggle={false}
      vertical={false}
    >
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="success"
      >
        Search
      </Button>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="danger"
      >
        Reset
      </Button>
    </ButtonGroup>
  </td>
</tr>
`;
